File: nodejs/src/repeatedfield.cc
// Protocol Buffers - Google's data interchange format
// Copyright 2015 Google Inc.  All rights reserved.
// https://developers.google.com/protocol-buffers/
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#include <nan.h>
#include "defs.h"
#include "message.h"
#include "repeatedfield.h"
#include "util.h"

#include <assert.h>
#include <vector>
#include <algorithm>

using namespace v8;
using namespace node;

namespace protobuf_js {

Persistent<Function> RepeatedField::constructor;
Persistent<Value> RepeatedField::prototype;

JS_OBJECT_INIT(RepeatedField);

void RepeatedField::Init(Handle<Object> exports) {
  NanEscapableScope();
  Local<FunctionTemplate> tpl = NanNew<FunctionTemplate>(New);
  tpl->SetClassName(NanNew<String>("RepeatedField"));
  tpl->InstanceTemplate()->SetInternalFieldCount(JS_OBJECT_WRAP_SLOTS);

  tpl->InstanceTemplate()->SetAccessor(
      NanNew("length"), LengthGetter);
  tpl->InstanceTemplate()->SetAccessor(
      NanNew("type"), TypeGetter);
  tpl->InstanceTemplate()->SetAccessor(
      NanNew("subdesc"), SubDescGetter);
  tpl->InstanceTemplate()->SetIndexedPropertyHandler(
      IndexGetter, IndexSetter, 0, IndexDeleter, 0);

#define M(jsname, cppname)                                   \
  tpl->PrototypeTemplate()->Set(                             \
      NanNew<String>(#jsname),                               \
      NanNew<FunctionTemplate>(cppname)->GetFunction())

  M(pop, Pop);
  M(push, Push);
  M(shift, Shift);
  M(unshift, Unshift);
  M(toString, ToString);
  M(resize, Resize);
  M(newEmpty, NewEmpty);

#undef M

  NanAssignPersistent(constructor, tpl->GetFunction());
  // Construct an instance in order to get the prototype object.
  Local<Value> arg = NanNew<Integer>(UPB_TYPE_INT32);
  NanAssignPersistent(prototype,
                      NanNew(constructor)->NewInstance(1, &arg)->GetPrototype());

  exports->Set(NanNew<String>("RepeatedField"), NanNew(constructor));
}

RepeatedField::RepeatedField() {
  // Defaults: will be replaced by HandleCtorArgs() after object is constructed
  // and attached to the JS object.
  type_ = UPB_TYPE_INT32;
  submsg_ = NULL;
  subenum_ = NULL;
}

NAN_METHOD(RepeatedField::New) {
  NanEscapableScope();
  if (!args.IsConstructCall()) {
    NanThrowError("Not called as constructor");
    NanReturnUndefined();
  }

  RepeatedField* self = new RepeatedField();
  self->Wrap<RepeatedField>(args.This());
  if (!self->HandleCtorArgs(args)) {
    NanReturnUndefined();
  }
  NanReturnValue(args.This());
}

bool RepeatedField::HandleCtorArgs(_NAN_METHOD_ARGS_TYPE args) {
  NanEscapableScope();

  if (args.Length() == 0) {
    NanThrowError("Expected at least one arg to RepeatedField constructor "
                  "(field type, or field type and message class)");
    return false;
  } else if (args.Length() >= 1 && args.Length() <= 3) {
    // One-arg form:   (FieldDescriptor.TYPE_***).
    // Two-arg form:   (FieldDescriptor.TYPE_MESSAGE, MessageClass/Desc).
    // Two-arg form:   (FieldDescriptor.TYPE_***, init_array).
    // Three-arg form: (FieldDescriptor.TYPE_MESSAGE, MessageClass/Desc,
    //                  init_array).
    Local<Value> type_value = NanNew(args[0]);
    if (!FieldDescriptor::ParseTypeValue(type_value, &type_)) {
      return false;
    }

    if (type_ == UPB_TYPE_MESSAGE && args.Length() > 1) {
      Local<Value> msgclass_or_desc = NanNew(args[1]);
      Local<Object> descriptor;
      if (msgclass_or_desc->IsFunction()) {
        Local<Function> msgclass_fn = msgclass_or_desc.As<Function>();
        Local<Value> descriptorval =
            NanNew(msgclass_fn->Get(NanNew<String>("descriptor")));
        if (!descriptorval->IsObject()) {
          NanThrowError("No descriptor property on message class");
          return false;
        }
        descriptor = NanNew(descriptorval->ToObject());
      } else if (msgclass_or_desc->IsObject()) {
        descriptor = msgclass_or_desc->ToObject();
      } else {
        NanThrowError("Expected message class or descriptor as second "
                      "argument to RepeatedField constructor");
        return false;
      }

      if (descriptor->GetPrototype() != Descriptor::prototype) {
        NanThrowError("Invalid descriptor object");
        return false;
      }

      submsg_ = Descriptor::unwrap(descriptor);
    } else {
      submsg_ = NULL;
    }

    if (type_ == UPB_TYPE_ENUM && args.Length() > 1) {
      if (!args[1]->IsObject()) {
        NanThrowError("Expected EnumDescriptor or enum object");
        return false;
      }
      Local<Object> enumobj_or_desc = NanNew(args[1]->ToObject());
      Local<Object> descriptor;
      if (enumobj_or_desc->GetPrototype() == EnumDescriptor::prototype) {
        descriptor = enumobj_or_desc;
      } else {
        Local<Value> descriptor_prop = enumobj_or_desc->Get(
            NanNew<String>("descriptor"));
        if (descriptor_prop->IsObject() &&
            descriptor_prop->ToObject()->GetPrototype() ==
            EnumDescriptor::prototype) {
          descriptor = descriptor_prop->ToObject();
        }
      }

      if (descriptor.IsEmpty()) {
        NanThrowError("Expected enum object or descriptor");
        return false;
      }

      subenum_ = EnumDescriptor::unwrap(descriptor);
    }

    if (type_ == UPB_TYPE_MESSAGE && !submsg_) {
      NanThrowError("RepeatedField created with message type but no submsg");
      return false;
    }
    if (type_ == UPB_TYPE_ENUM && !subenum_) {
      NanThrowError("RepeatedField created with enum type but no subenum");
      return false;
    }

    int init_array_idx = (submsg_ != NULL || subenum_ != NULL) ? 2 : 1;
    if (args.Length() > init_array_idx) {
      Local<Value> init_array_val = NanNew(args[init_array_idx]);
      if (!init_array_val->IsArray()) {
        NanThrowError(
            "Expected array as last arg to RepeatedField constructor");
        return false;
      }
      Local<Array> init_array = init_array_val.As<Array>();

      for (int i = 0; i < init_array->Length(); i++) {
        if (!DoPush(init_array->Get(i), /* allow_copy = */ false)) {
          return false;
        }
      }
    }

    return true;
  } else {
    NanThrowError("Too many args to RepeatedField constructor");
    return false;
  }
}

NAN_GETTER(RepeatedField::LengthGetter) {
  NanEscapableScope();
  RepeatedField* self = unwrap(args.This());
  if (!self) {
    NanReturnUndefined();
  }
  NanReturnValue(NanNew<Integer>(self->values_.size()));
}

NAN_INDEX_GETTER(RepeatedField::IndexGetter) {
  NanEscapableScope();
  RepeatedField* self = unwrap(args.This());
  if (!self) {
    NanReturnUndefined();
  }
  if (index < self->values_.size()) {
    NanReturnValue(NanNew(self->values_[index]));
  } else {
    // Out-of-bounds access
    NanReturnUndefined();
  }
}

NAN_INDEX_SETTER(RepeatedField::IndexSetter) {
  NanEscapableScope();
  RepeatedField* self = unwrap(args.This());
  if (!self) {
    NanReturnUndefined();
  }

  if (index >= self->values_.size()) {
    NanThrowError("Out-of-bounds assignment to repeated field");
    NanReturnUndefined();
  }

  Local<Value> converted_val =
      NanNew(ProtoMessage::CheckConvertElement(self->type_,
                                               self->submsg_,
                                               value,
                                               /* allow_null = */ false,
                                               /* allow_copy = */ false));
  if (converted_val.IsEmpty() || converted_val->IsUndefined()) {
    NanReturnUndefined();
  }

  NanAssignPersistent(self->values_[index], converted_val);
  NanReturnValue(converted_val);
}

NAN_INDEX_DELETER(RepeatedField::IndexDeleter) {
  NanEscapableScope();
  NanThrowError("Delete not supported on a repeated field element");
  NanReturnValue(NanTrue());  // true == intercepted request
}

NAN_METHOD(RepeatedField::Pop) {
  NanEscapableScope();
  RepeatedField* self = unwrap(args.This());
  if (!self) {
    NanReturnUndefined();
  }
  if (self->values_.empty()) {
    NanReturnUndefined();
  } else {
    Local<Value> last_obj = NanNew(self->values_.back());
    self->values_.pop_back();
    NanReturnValue(last_obj);
  }
}

bool RepeatedField::DoPush(Local<Value> value, bool allow_copy) {
  NanEscapableScope();
  if (value->IsNull()) {
    NanThrowError("Cannot set a value to null in a repeated field");
    return false;
  }
  Local<Value> converted = NanNew(
      ProtoMessage::CheckConvertElement(type_, submsg_, value,
                                        /* allow_null = */ false,
                                        allow_copy));
  if (converted.IsEmpty() || converted->IsUndefined()) {
    return false;
  }

  values_.resize(values_.size() + 1);
  NanAssignPersistent(values_.back(), converted);
  return true;
}

NAN_METHOD(RepeatedField::Push) {
  NanEscapableScope();
  RepeatedField* self = unwrap(args.This());
  if (!self) {
    NanReturnUndefined();
  }
  if (args.Length() != 1) {
    NanThrowError("Push expects one argument");
    NanReturnUndefined();
  }
  if (!self->DoPush(args[0], /* allow_copy = */ false)) {
    NanReturnUndefined();
  }
  NanReturnValue(args[0]);
}

NAN_METHOD(RepeatedField::Shift) {
  NanEscapableScope();
  RepeatedField* self = unwrap(args.This());
  if (!self) {
    NanReturnUndefined();
  }
  CheckArgs checkargs(args);
  if (!checkargs.ok()) {
    NanReturnUndefined();
  }

  if (self->values_.empty()) {
    NanReturnUndefined();
  }

  Local<Value> front_val = NanNew(self->values_.front());
  for (int i = 0; i < self->values_.size() - 1; i++) {
    NanAssignPersistent(self->values_[i],
            NanNew(self->values_[i + 1]));
  }
  self->values_.resize(self->values_.size() - 1);

  NanReturnValue(front_val);
}

NAN_METHOD(RepeatedField::Unshift) {
  NanEscapableScope();
  RepeatedField* self = unwrap(args.This());
  if (!self) {
    NanReturnUndefined();
  }
  if (args.Length() != 1) {
    NanThrowError("Expected one argument to RepeatedField.unshift");
    NanReturnUndefined();
  }

  Local<Value> value = args[0];

  if (value->IsNull()) {
    NanThrowError("Cannot set a value to null in a repeated field");
    NanReturnUndefined();
  }
  Local<Value> converted = NanNew(
      ProtoMessage::CheckConvertElement(self->type_, self->submsg_, value,
                                        /* allow_null = */ false,
                                        /* allow_copy = */ false));
  if (converted.IsEmpty() || converted->IsUndefined()) {
    NanReturnUndefined();
  }

  self->values_.resize(self->values_.size() + 1);
  for (int i = self->values_.size() - 1; i > 0; --i) {
    NanAssignPersistent(self->values_[i], NanNew(self->values_[i - 1]));
  }
  NanAssignPersistent(self->values_[0], converted);

  NanReturnValue(converted);
}

NAN_METHOD(RepeatedField::ToString) {
  NanEscapableScope();
  RepeatedField* self = unwrap(args.This());
  if (!self) {
    NanReturnUndefined();
  }
  CheckArgs checkargs(args);
  if (!checkargs.ok()) {
    NanReturnUndefined();
  }

  Local<Object> type_desc;
  if (self->type_ == UPB_TYPE_MESSAGE) {
    type_desc = NanNew(self->submsg_->object());
  } else if (self->type_ == UPB_TYPE_ENUM) {
    type_desc = NanNew(self->subenum_->object());
  }

  std::string value = "[";
  bool first = true;
  for (int i = 0; i < self->values_.size(); i++) {
    if (first) {
      first = false;
    } else {
      value += ", ";
    }
    value += ProtoMessage::ElementString(self->type_,
                                         type_desc,
                                         NanNew(self->values_[i]));
  }
  value += "]";

  NanReturnValue(NanNew<String>(value.data(), value.size()));
}

NAN_METHOD(RepeatedField::Resize) {
  NanEscapableScope();
  RepeatedField* self = RepeatedField::unwrap(args.This());
  if (!self) {
    NanReturnUndefined();
  }

  if (args.Length() != 1 || !args[0]->IsInt32()) {
    NanThrowError("Expected one integer argument");
    NanReturnUndefined();
  }

  size_t newsize = args[0]->Int32Value();
  size_t oldsize = self->values_.size();
  self->values_.resize(newsize);
  if (newsize > oldsize) {
    for (size_t i = oldsize; i < newsize; i++) {
      NanEscapableScope();
      Local<Value> new_element;
      if (self->type_ == UPB_TYPE_MESSAGE) {
        new_element =
            NanNew(self->submsg_->Constructor()->NewInstance(0, NULL));
      } else {
        new_element = NanNew(ProtoMessage::NewElement(self->type_));
      }
      NanAssignPersistent(self->values_[i], new_element);
    }
  }

  NanReturnUndefined();
}

NAN_METHOD(RepeatedField::NewEmpty) {
  NanEscapableScope();
  RepeatedField* self = RepeatedField::unwrap(args.This());
  if (!self) {
    NanReturnUndefined();
  }

  if (args.Length() > 0) {
    NanThrowError("Expected no arguments");
    NanReturnUndefined();
  }

  Handle<Value> argv[2];
  int argc = 1;

  argv[0] = NanNew<Integer>(self->type_);
  if (self->type_ == UPB_TYPE_MESSAGE) {
    argv[1] = NanNew(self->submsg_->object());
    argc = 2;
  } else if (self->type_ == UPB_TYPE_ENUM) {
    argv[1] = NanNew(self->subenum_->object());
    argc = 2;
  }

  Local<Value> new_rptfield =
      NanNew(NanNew(constructor)->NewInstance(argc, argv));
  NanReturnValue(new_rptfield);
}

NAN_GETTER(RepeatedField::TypeGetter) {
  NanEscapableScope();
  RepeatedField* self = RepeatedField::unwrap(args.This());
  if (!self) {
    NanReturnUndefined();
  }
  NanReturnValue(NanNew<Integer>(self->type_));
}

NAN_GETTER(RepeatedField::SubDescGetter) {
  NanEscapableScope();
  RepeatedField* self = RepeatedField::unwrap(args.This());
  if (!self) {
    NanReturnUndefined();
  }
  if (self->type_ == UPB_TYPE_MESSAGE) {
    NanReturnValue(NanNew(self->submsg_->object()));
  } else if (self->type_ == UPB_TYPE_ENUM) {
    NanReturnValue(NanNew(self->subenum_->object()));
  } else {
    NanReturnUndefined();
  }
}

}  // namespace protobuf_js

File: csharp/src/ProtocolBuffers/Collections/RepeatedField.cs
#region Copyright notice and license
// Protocol Buffers - Google's data interchange format
// Copyright 2015 Google Inc.  All rights reserved.
// https://developers.google.com/protocol-buffers/
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#endregion
    
using System;
using System.Collections;
using System.Collections.Generic;

namespace Google.Protobuf.Collections
{
    /// <summary>
    /// The contents of a repeated field: essentially, a collection with some extra
    /// restrictions (no null values) and capabilities (deep cloning and freezing).
    /// </summary>
    /// <typeparam name="T">The element type of the repeated field.</typeparam>
    public sealed class RepeatedField<T> : IList<T>, IList, IDeepCloneable<RepeatedField<T>>, IEquatable<RepeatedField<T>>, IFreezable
    {
        private static readonly T[] EmptyArray = new T[0];

        private bool frozen;
        private const int MinArraySize = 8;
        private T[] array = EmptyArray;
        private int count = 0;

        /// <summary>
        /// Creates a deep clone of this repeated field.
        /// </summary>
        /// <remarks>
        /// If the field type is
        /// a message type, each element is also cloned; otherwise, it is
        /// assumed that the field type is primitive (including string and
        /// bytes, both of which are immutable) and so a simple copy is
        /// equivalent to a deep clone.
        /// </remarks>
        /// <returns>A deep clone of this repeated field.</returns>
        public RepeatedField<T> Clone()
        {
            RepeatedField<T> clone = new RepeatedField<T>();
            // Clone is implicitly *not* frozen, even if this object is.
            if (array != EmptyArray)
            {
                clone.array = (T[])array.Clone();
                IDeepCloneable<T>[] cloneableArray = clone.array as IDeepCloneable<T>[];
                if (cloneableArray != null)
                {
                    for (int i = 0; i < count; i++)
                    {
                        clone.array[i] = cloneableArray[i].Clone();
                    }
                }
            }
            clone.count = count;
            return clone;
        }

        public void AddEntriesFrom(CodedInputStream input, FieldCodec<T> codec)
        {
            // TODO: Inline some of the Add code, so we can avoid checking the size on every
            // iteration and the mutability.
            uint tag = input.LastTag;
            var reader = codec.ValueReader;
            // Value types can be packed or not.
            if (typeof(T).IsValueType && WireFormat.GetTagWireType(tag) == WireFormat.WireType.LengthDelimited)
            {
                int length = input.ReadLength();
                if (length > 0)
                {
                    int oldLimit = input.PushLimit(length);
                    while (!input.ReachedLimit)
                    {
                        Add(reader(input));
                    }
                    input.PopLimit(oldLimit);
                }
                // Empty packed field. Odd, but valid - just ignore.
            }
            else
            {
                // Not packed... (possibly not packable)
                do
                {
                    Add(reader(input));
                } while (input.MaybeConsumeTag(tag));
            }
        }

        public int CalculateSize(FieldCodec<T> codec)
        {
            if (count == 0)
            {
                return 0;
            }
            uint tag = codec.Tag;
            if (typeof(T).IsValueType && WireFormat.GetTagWireType(tag) == WireFormat.WireType.LengthDelimited)
            {
                int dataSize = CalculatePackedDataSize(codec);
                return CodedOutputStream.ComputeRawVarint32Size(tag) +
                    CodedOutputStream.ComputeRawVarint32Size((uint)dataSize) +
                    dataSize;
            }
            else
            {
                var sizeCalculator = codec.ValueSizeCalculator;
                int size = count * CodedOutputStream.ComputeRawVarint32Size(tag);
                for (int i = 0; i < count; i++)
                {
                    size += sizeCalculator(array[i]);
                }
                return size;
            }
        }

        private int CalculatePackedDataSize(FieldCodec<T> codec)
        {
            int fixedSize = codec.FixedSize;
            if (fixedSize == 0)
            {
                var calculator = codec.ValueSizeCalculator;
                int tmp = 0;
                for (int i = 0; i < count; i++)
                {
                    tmp += calculator(array[i]);
                }
                return tmp;
            }
            else
            {
                return fixedSize * Count;
            }
        }

        public void WriteTo(CodedOutputStream output, FieldCodec<T> codec)
        {
            if (count == 0)
            {
                return;
            }
            var writer = codec.ValueWriter;
            var tag = codec.Tag;
            if (typeof(T).IsValueType && WireFormat.GetTagWireType(tag) == WireFormat.WireType.LengthDelimited)
            {
                // Packed primitive type
                uint size = (uint)CalculatePackedDataSize(codec);
                output.WriteTag(tag);
                output.WriteRawVarint32(size);
                for (int i = 0; i < count; i++)
                {
                    writer(output, array[i]);
                }
            }
            else
            {
                // Not packed: a simple tag/value pair for each value.
                // Can't use codec.WriteTagAndValue, as that omits default values.
                for (int i = 0; i < count; i++)
                {
                    output.WriteTag(tag);
                    writer(output, array[i]);
                }
            }
        }

        public bool IsFrozen { get { return frozen; } }

        public void Freeze()
        {
            frozen = true;
            IFreezable[] freezableArray = array as IFreezable[];
            if (freezableArray != null)
            {
                for (int i = 0; i < count; i++)
                {
                    freezableArray[i].Freeze();
                }
            }
        }

        private void EnsureSize(int size)
        {
            if (array.Length < size)
            {
                size = Math.Max(size, MinArraySize);
                int newSize = Math.Max(array.Length * 2, size);
                var tmp = new T[newSize];
                Array.Copy(array, 0, tmp, 0, array.Length);
                array = tmp;
            }
        }

        public void Add(T item)
        {
            if (item == null)
            {
                throw new ArgumentNullException("item");
            }
            this.CheckMutable();
            EnsureSize(count + 1);
            array[count++] = item;
        }

        public void Clear()
        {
            this.CheckMutable();
            array = EmptyArray;
            count = 0;
        }

        public bool Contains(T item)
        {
            return IndexOf(item) != -1;
        }

        public void CopyTo(T[] array, int arrayIndex)
        {
            Array.Copy(this.array, 0, array, arrayIndex, count);
        }

        public bool Remove(T item)
        {
            this.CheckMutable();
            int index = IndexOf(item);
            if (index == -1)
            {
                return false;
            }            
            Array.Copy(array, index + 1, array, index, count - index - 1);
            count--;
            array[count] = default(T);
            return true;
        }

        public int Count { get { return count; } }

        // TODO(jonskeet): If we implement freezing, make this reflect it.
        public bool IsReadOnly { get { return IsFrozen; } }

        public void Add(RepeatedField<T> values)
        {
            if (values == null)
            {
                throw new ArgumentNullException("values");
            }
            this.CheckMutable();
            EnsureSize(count + values.count);
            // We know that all the values will be valid, because it's a RepeatedField.
            Array.Copy(values.array, 0, array, count, values.count);
            count += values.count;
        }

        public void Add(IEnumerable<T> values)
        {
            if (values == null)
            {
                throw new ArgumentNullException("values");
            }
            this.CheckMutable();
            // TODO: Check for ICollection and get the Count?
            foreach (T item in values)
            {
                Add(item);
            }
        }

        public RepeatedField<T>.Enumerator GetEnumerator()
        {
            return new Enumerator(this);
        }

        IEnumerator<T> IEnumerable<T>.GetEnumerator()
        {
            return GetEnumerator();
        }

        public override bool Equals(object obj)
        {
            return Equals(obj as RepeatedField<T>);
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }        

        public override int GetHashCode()
        {
            int hash = 0;
            for (int i = 0; i < count; i++)
            {
                hash = hash * 31 + array[i].GetHashCode();
            }
            return hash;
        }

        public bool Equals(RepeatedField<T> other)
        {
            if (ReferenceEquals(other, null))
            {
                return false;
            }
            if (ReferenceEquals(other, this))
            {
                return true;
            }
            if (other.Count != this.Count)
            {
                return false;
            }
            // TODO(jonskeet): Does this box for enums?
            EqualityComparer<T> comparer = EqualityComparer<T>.Default;
            for (int i = 0; i < count; i++)
            {
                if (!comparer.Equals(array[i], other.array[i]))
                {
                    return false;
                }
            }
            return true;
        }

        public int IndexOf(T item)
        {
            if (item == null)
            {
                throw new ArgumentNullException("item");
            }
            // TODO(jonskeet): Does this box for enums?
            EqualityComparer<T> comparer = EqualityComparer<T>.Default;
            for (int i = 0; i < count; i++)
            {
                if (comparer.Equals(array[i], item))
                {
                    return i;
                }
            }
            return -1;
        }

        public void Insert(int index, T item)
        {
            if (item == null)
            {
                throw new ArgumentNullException("item");
            }
            if (index < 0 || index > count)
            {
                throw new ArgumentOutOfRangeException("index");
            }
            this.CheckMutable();
            EnsureSize(count + 1);
            Array.Copy(array, index, array, index + 1, count - index);
            array[index] = item;
            count++;
        }

        public void RemoveAt(int index)
        {
            if (index < 0 || index >= count)
            {
                throw new ArgumentOutOfRangeException("index");
            }
            this.CheckMutable();
            Array.Copy(array, index + 1, array, index, count - index - 1);
            count--;
            array[count] = default(T);
        }

        public T this[int index]
        {
            get
            {
                if (index < 0 || index >= count)
                {
                    throw new ArgumentOutOfRangeException("index");
                }
                return array[index];
            }
            set
            {
                if (index < 0 || index >= count)
                {
                    throw new ArgumentOutOfRangeException("index");
                }
                this.CheckMutable();
                if (value == null)
                {
                    throw new ArgumentNullException("value");
                }
                array[index] = value;
            }
        }

        #region Explicit interface implementation for IList and ICollection.
        bool IList.IsFixedSize { get { return IsFrozen; } }

        void ICollection.CopyTo(Array array, int index)
        {
            Array.Copy(this.array, 0, array, index, count);
        }

        bool ICollection.IsSynchronized { get { return false; } }

        object ICollection.SyncRoot { get { return this; } }

        object IList.this[int index]
        {
            get { return this[index]; }
            set { this[index] = (T)value; }
        }

        int IList.Add(object value)
        {
            Add((T) value);
            return count - 1;
        }

        bool IList.Contains(object value)
        {
            return (value is T && Contains((T)value));
        }

        int IList.IndexOf(object value)
        {
            if (!(value is T))
            {
                return -1;
            }
            return IndexOf((T)value);
        }

        void IList.Insert(int index, object value)
        {
            Insert(index, (T) value);
        }

        void IList.Remove(object value)
        {
            if (!(value is T))
            {
                return;
            }
            Remove((T)value);
        }
        #endregion

        public struct Enumerator : IEnumerator<T>
        {
            private int index;
            private readonly RepeatedField<T> field;

            public Enumerator(RepeatedField<T> field)
            {
                this.field = field;
                this.index = -1;
            }

            public bool MoveNext()
            {
                if (index + 1 >= field.Count)
                {
                    index = field.Count;
                    return false;
                }
                index++;
                return true;
            }

            public void Reset()
            {
                index = -1;
            }

            public T Current
            {
                get
                {
                    if (index == -1 || index >= field.count)
                    {
                        throw new InvalidOperationException();
                    }
                    return field.array[index];
                }
            }

            object IEnumerator.Current
            {
                get { return Current; }
            }

            public void Dispose()
            {
            }
        }
    }
}
